src/table.js

```diff
@@ -542,15 +550,81 @@ export function getTypeValidator(colType) {
   }
 }
 
+export function coerceToType(value, type) {
+  switch (type) {
+    case "string":
+      return typeof value === "string" || value == null
+        ? value
+        : String(value);
+    case "boolean":
+      if (typeof value === "string") {
+        const trimValue = value.trim();
+        return trimValue === "true" ? true : trimValue === "false" ? false : null;
+      }
+      return typeof value === "boolean" || value == null
+        ? value
+        : Boolean(value);
+    case "bigint":
+      return typeof value === "bigint"
+        ? value
+        : value === 0 || value === true || value === false
+        ? // eslint-disable-next-line no-undef
+          BigInt(value)
+        : !value || isNaN(value) || !Number.isInteger(+value)
+        ? NaN
+        : // eslint-disable-next-line no-undef
+          BigInt(value);
+    case "integer": // not a target type for coercion, but can be inferred
+    case "number": {
+      return typeof value === "number"
+        ? value
+        : value == null || value === ""
+        ? NaN
+        : Number(value);
+    }
+    case "date": {
+      if (value instanceof Date) return value;
+      const trimValue = typeof value === "string" ? value.trim() : value;
+      return new Date(trimValue);
+    }
+    case "array":
+    case "object":
+    case "buffer":
+    case "other":
+      return value;
+    default:
+      throw new Error(`Unable to coerce to type: ${type}`);
+  }
+}
+
 // This function applies table cell operations to an in-memory table (array of
 // objects); it should be equivalent to the corresponding SQL query. TODO Use
 // DuckDBClient for data arrays, too, and then we wouldn’t need our own __table
 // function to do table operations on in-memory data?
 export function __table(source, operations) {
   const input = source;
   let {schema, columns} = source;
-  let primitive = arrayIsPrimitive(source);
-  if (primitive) source = Array.from(source, (value) => ({value}));
+  let inferredSchema = false;
+  if (!isQueryResultSetSchema(schema)) {
+    schema = inferSchema(source, columns);
+    inferredSchema = true;
+  }
+  // Combine column types from schema with user-selected types in operations
+  const types = new Map(schema.map(({name, type}) => [name, type]));
+  if (operations.type) {
+    for (const {name, type} of operations.type) {
+      types.set(name, type);
+      // update schema with user-selected type
+      if (schema === input.schema) schema = schema.slice(); // copy on write
+      const colIndex = schema.findIndex((col) => col.name === name);
+      if (colIndex > -1) schema[colIndex] = {...schema[colIndex], type};
```

Comment:
[optional] As an alternative to generating `{name, type, inferred}` in `inferSchema`, we could preserve the original type here when applying type coercion:

```suggestion
      if (colIndex > -1) schema[colIndex] = {name, originalType: schema[colIndex].type, type};
```

That might be useful if we want to preserve the original type when applying type coercion even in the case where the schema is not inferred. I don’t recall if we ended up allowing type coercion/assertion in the case where a schema is already explicitly provided by the data source though. But since you’ve already done some downstream work using this schema feel free to ignore this suggestion if it’s not helpful!

Reply:
This is a good thought, but I think it makes sense to keep the association with `inferred`, at least for now, because the downstream use for it is restricted to sources we've inferred a schema for.